src/Compilers/CSharp/Test/Semantic/Semantics/RefExtensionMethodsTests.cs

```diff
@@ -0,0 +1,1992 @@
+        }
+
+        [Fact]
+        public void ExtensionMethods_RValues_RefReadOnly_Allowed()
+        {
+            var code = @"
+public static class Extensions
+{
+    public static void PrintValue(ref readonly this int p)
+    {
+        System.Console.Write(p);
+    }
+}
+public static class Program
+{
+    public static void Main()
+    {
+        5.PrintValue();
+        Extensions.PrintValue(5);
+    }
+}";
+
+            CompileAndVerify(code, additionalRefs: new[] { SystemCoreRef }, expectedOutput: "55");
+        }
+
+        [Fact]
+        public void ExtensionMethods_LValues_RefReadOnly_Allowed()
+        {
+            var code = @"
+public static class Extensions
+{
+    public static void PrintValue(ref readonly this int p)
+    {
+        System.Console.Write(p);
+    }
+}
+public static class Program
+{
+    public static void Main()
+    {
+        int x = 5;
+        x.PrintValue();
+        Extensions.PrintValue(x);
+    }
+}";
+
+            CompileAndVerify(code, additionalRefs: new[] { SystemCoreRef }, expectedOutput: "55");
+        }
+
+        [Fact]
+        public void ExtensionMethods_NullConditionalOperator_Ref_NotAllowed()
+        {
+            var code = @"
+public struct TestType
+{
+    public int GetValue() => 0;
+}
+public static class Extensions
+{
+    public static void Call(ref this TestType obj)
+    {
+        var value1 = obj?.GetValue();        // This should be an error
+        var value2 = obj.GetValue();         // This should be OK
+    }
+}";
+
+            CreateCompilationWithMscorlibAndSystemCore(code).VerifyDiagnostics(
+                // (10,25): error CS0023: Operator '?' cannot be applied to operand of type 'TestType'
+                //         var value1 = obj?.GetValue();        // This should be an error
+                Diagnostic(ErrorCode.ERR_BadUnaryOp, "?").WithArguments("?", "TestType").WithLocation(10, 25));
+        }
+
+        [Fact]
+        public void ExtensionMethods_NullConditionalOperator_RefReadOnly_NotAllowed()
+        {
+            var code = @"
+public struct TestType
+{
+    public int GetValue() => 0;
+}
+public static class Extensions
+{
+    public static void Call(ref readonly this TestType obj)
+    {
+        var value1 = obj?.GetValue();        // This should be an error
+        var value2 = obj.GetValue();         // This should be OK
+    }
+}";
+
+            CreateCompilationWithMscorlibAndSystemCore(code).VerifyDiagnostics(
+                // (10,25): error CS0023: Operator '?' cannot be applied to operand of type 'TestType'
+                //         var value1 = obj?.GetValue();        // This should be an error
+                Diagnostic(ErrorCode.ERR_BadUnaryOp, "?").WithArguments("?", "TestType").WithLocation(10, 25));
+        }
+
+        [Fact]
+        public void RefExtensionMethodsReceiverTypes_ValueTypes_Allowed()
+        {
+            var reference = CreateCompilationWithMscorlibAndSystemCore(@"
```

Comment:
Because of `ExtensionAttribute`. It exists in `System.Core` unfortunately.

---
In reply to: [122884997](https://github.com/dotnet/roslyn/pull/19703#discussion_r122884997) [](ancestors = 122884997)